File: src/main/java/com/timeback/device/access/AccessGate.java
package com.timeback.device.access;

import com.timeback.device.contract.AccessState;
import com.timeback.device.contract.SettingsOpenResult;
import com.timeback.device.contract.TimeSource;
import com.timeback.device.contract.UsageAccessGateway;
import com.timeback.device.contract.UsageAccessStatus;

public final class AccessGate {
    private final UsageAccessGateway gateway;
    private final TimeSource timeSource;

    public AccessGate(UsageAccessGateway gateway, TimeSource timeSource) {
        this.gateway = gateway;
        this.timeSource = timeSource;
    }

    public AccessState readAccessState() {
        long observedAt = timeSource.nowMillis();
        try {
            UsageAccessStatus status = gateway.readCurrentStatus();
            if (status == UsageAccessStatus.GRANTED) {
                return new AccessState.Granted(observedAt);
            }
            return new AccessState.Blocked(observedAt);
        } catch (RuntimeException error) {
            return new AccessState.Failure(
                    observedAt,
                    error.getMessage() == null ? "usage access status unavailable" : error.getMessage()
            );
        }
    }

    public SettingsOpenResult openUsageAccessSettings() {
        try {
            if (gateway.openUsageAccessSettings()) {
                return SettingsOpenResult.Opened.INSTANCE;
            }
            return new SettingsOpenResult.Failure("usage access settings could not be opened");
        } catch (RuntimeException error) {
            return new SettingsOpenResult.Failure(
                    error.getMessage() == null ? "usage access settings unavailable" : error.getMessage()
            );
        }
    }
}

File: src/main/java/com/timeback/device/access/AccessGate.kt
package com.timeback.device.access

import com.timeback.device.contract.AccessState
import com.timeback.device.contract.SettingsOpenResult
import com.timeback.device.contract.TimeSource
import com.timeback.device.contract.UsageAccessGateway
import com.timeback.device.contract.UsageAccessStatus

class AccessGate(
    private val gateway: UsageAccessGateway,
    private val timeSource: TimeSource,
) {
    fun readAccessState(): AccessState {
        val observedAt = timeSource.nowMillis()
        return try {
            when (gateway.readCurrentStatus()) {
                UsageAccessStatus.GRANTED -> AccessState.Granted(observedAt)
                UsageAccessStatus.NOT_GRANTED -> AccessState.Blocked(observedAt)
            }
        } catch (error: RuntimeException) {
            AccessState.Failure(observedAt, error.message ?: "usage access status unavailable")
        }
    }

    fun openUsageAccessSettings(): SettingsOpenResult = try {
        if (gateway.openUsageAccessSettings()) {
            SettingsOpenResult.Opened
        } else {
            SettingsOpenResult.Failure("usage access settings could not be opened")
        }
    } catch (error: RuntimeException) {
        SettingsOpenResult.Failure(error.message ?: "usage access settings unavailable")
    }
}
